### PR TITLE
Add annotations and pieces tables

### PR DESCRIPTION
**Pull request by jaylorch**
This pull request creates the annotations table (for storing per-team, per-puzzle annotations for puzzles that need them) and the pieces table (for storing pieces of a puzzle that teams earn as they make progress).  It also creates CRUD pages for the pieces table.

### CODE REVIEW

File: Data/DataModel/Piece.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+	/// revealed to that team
+        /// </summary>
+        [Required]
+        public int ProgressLevel { get; set; }
+
+        /// <summary>
+        /// The contents of the piece
+        /// </summary>
+        [Required]
+        [StringLength(4096)]
```
Comment (morganbr) [inline Data/DataModel/Piece.cs]: Is there a reason for this limit?

File: Data/Migrations/20190219002519_AddAnnotationTable.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+                },
+                constraints: table =>
+                {
+                    table.PrimaryKey("PK_Annotations", x => new { x.PuzzleID, x.TeamID, x.Key });
+                    table.ForeignKey(
+                        name: "FK_Annotations_Puzzles_PuzzleID",
+                        column: x => x.PuzzleID,
+                        principalTable: "Puzzles",
+                        principalColumn: "ID",
+                        onDelete: ReferentialAction.Cascade);
```
Comment (morganbr) [inline Data/Migrations/20190219002519_AddAnnotationTable.cs]: Having cascading deletes for both Puzzle and Team is going to break when it meets up with #243 (which I'll be merging tonight) . You'll need to disable cascading deletes for one of them and manually delete annotations for it. See #243 for examples with PuzzleStatePerTeam.

File: Data/DataModel/Annotation.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Text;
+
+namespace ServerCore.DataModel
+{
+    public class Annotation
```
Comment (morganbr) [inline Data/DataModel/Annotation.cs]: Please add a comment explaining what an annotation is

File: Data/PuzzleServerContext.cs [hunk 4]
```diff
@@ -58,6 +60,7 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
             modelBuilder.Entity<PuzzleStatePerTeam>().HasKey(state => new { state.PuzzleID, state.TeamID });
             modelBuilder.Entity<HintStatePerTeam>().HasKey(state => new { state.TeamID, state.HintID });
             modelBuilder.Entity<Event>().HasIndex(eventObj => new { eventObj.UrlString }).IsUnique();
+	    modelBuilder.Entity<Annotation>().HasKey(state => new { state.PuzzleID, state.TeamID, state.Key });
```
Comment (morganbr) [inline Data/PuzzleServerContext.cs]: Tab

File: ServerCore/Pages/Pieces/Create.cshtml.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+using Microsoft.AspNetCore.Identity;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.RazorPages;
+using Microsoft.EntityFrameworkCore;
+using ServerCore.DataModel;
+using ServerCore.ModelBases;
+
+namespace ServerCore.Pages.Pieces
+{
+    [Authorize(Policy = "IsAuthorOfPuzzle")]
```
Comment (morganbr) [inline ServerCore/Pages/Pieces/Create.cshtml.cs]: This should probably be IsEventAdminOrAuthorOfPuzzle -- we generally let event admins do anything.

File: ServerCore/Pages/Pieces/Delete.cshtml.cs [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Identity;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore;
+using ServerCore.DataModel;
+using ServerCore.ModelBases;
+
+namespace ServerCore.Pages.Pieces
+{
+    [Authorize(Policy = "IsAuthorOfPuzzle")]
```
Comment (morganbr) [inline ServerCore/Pages/Pieces/Delete.cshtml.cs]: Ditto

File: ServerCore/Pages/Pieces/Delete.cshtml.cs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        }
+
+        [BindProperty]
+        public Piece Piece { get; set; }
+        
+        public int PuzzleId { get; set; }
+
+        public async Task<IActionResult> OnGetAsync(int? id)
+        {
+            if (id == null)
```
Comment (morganbr) [inline ServerCore/Pages/Pieces/Delete.cshtml.cs]: Why have a nullable ID if we'd return 404 if it's null?

File: ServerCore/Pages/Pieces/Delete.cshtml.cs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+            if (Piece == null)
+            {
+                return NotFound();
+            }
+            return Page();
+        }
+
+        public async Task<IActionResult> OnPostAsync(int? id)
+        {
+            if (id == null)
```
Comment (morganbr) [inline ServerCore/Pages/Pieces/Delete.cshtml.cs]: Ditto

File: ServerCore/Pages/Pieces/Edit.cshtml.cs [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Identity;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore;
+using ServerCore.DataModel;
+using ServerCore.ModelBases;
+
+namespace ServerCore.Pages.Pieces
+{
+    [Authorize(Policy = "IsAuthorOfPuzzle")]
```
Comment (morganbr) [inline ServerCore/Pages/Pieces/Edit.cshtml.cs]: Same auth comment

File: ServerCore/ServerCore.csproj [hunk 10]
```diff
@@ -25,6 +25,11 @@
     <PackageReference Include="Microsoft.AspNetCore.All" Version="2.1.2" />
     <PackageReference Include="Microsoft.AspNetCore.Authentication.MicrosoftAccount" Version="2.1.2" />
     <PackageReference Include="Microsoft.EntityFrameworkCore.Proxies" Version="2.1.0" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="2.2.2" />
```
Comment (morganbr) [inline ServerCore/ServerCore.csproj]: Why did you add these?

File: Data/DataModel/Annotation.cs [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+        /// The version number, which increases with each change to the same annotation key
+        /// </summary>
+        [Required]
+        public int Version { get; set; }
+
+        /// <summary>
+        /// The annotation contents
+        /// </summary>
+        [Required]
+        [StringLength(255)]
```
[the remark this answers is not in the document]
Comment (jaylorch) [reply Data/DataModel/Annotation.cs]: Yes, it's long enough.  I set it intentionally small so that no one could use annotations to store huge amounts of data on our server.  Maybe some future puzzle writer will want it longer, but at that point it should be easy to just increase this number.

File: Data/DataModel/Piece.cs (hunk 1 above)
Comment (jaylorch) [reply Data/DataModel/Piece.cs]: Not really; I'll remove it.

File: Data/Migrations/20190219002519_AddAnnotationTable.cs (hunk 2 above)
Comment (jaylorch) [reply Data/Migrations/20190219002519_AddAnnotationTable.cs]: I don't understand.

File: ServerCore/ServerCore.csproj (hunk 10 above)
Comment (jaylorch) [reply ServerCore/ServerCore.csproj]: The wiki said to follow the instructions at 
https://docs.microsoft.com/en-us/ef/core/get-started/install/
to get started with EF Core.  Those instructions said to install Microsoft.EntityFrameworkCore.SqlServer and Microsoft.EntityFrameworkCore.Tools.  So I did.

I then updated the wiki here to include all those details:
https://github.com/PuzzleServer/mainpuzzleserver/wiki/Local-Development-Environment-Setup
